### PR TITLE
Userspace maps (+ update operation)

### PR DESCRIPTION
**Pull request by laarmen**
Sorry, this is a bit of a monster PR.

Apart from changing the Lookup trait to a LookupMut (to be able to add a non-mut variant), there's a tiny API break, as I changed the Result<Option<..>> into a simple Result<..> as the None variant matches the Error::InvalidMapName anyways.

I also introduced a Result<T> type alias in an effort to get the type names under control.

Oh, and we can now update maps in a safe way :)

### CODE REVIEW

File: rebpf/src/libbpf.rs [hunk 1]
```diff
@@ -356,17 +356,14 @@ pub fn bpf_object__find_program_by_title(
 }
 
 #[allow(non_snake_case)]
-pub fn bpf_object__find_map_by_name(
-    bpf_object: &BpfObject,
-    name: &str,
-) -> Result<Option<BpfMap>, Error> {
```
Comment (uccidibuti) [inline rebpf/src/libbpf.rs:342]: I don't know if this is correct, if I call a "find/lookup function" and no element was founded I expected a None as return value and no an Error. Isn't it the same case of map lookup function where we return None?

**Comment by uccidibuti** [comment]
Thank you for these upgrades and fixed my macro bugs, I have accept your PR with a comment about bpf_object__find_map_by_name, even if I know that removing Option return value makes client code a bit easier to write, I don't think that it is right, but it seem that you have a better Rust and bpf background then me so if you are convinced that this is right I can leave your code without any problems.. Let me know about  bpf_object__find_map_by_name..

File: rebpf/src/libbpf.rs (hunk 1 above)
Comment (laarmen) [reply rebpf/src/libbpf.rs:342]: Well, the way I see it is that an element not found is only a special case of error. If it were the only possible error, then I'd indeed go for the Option<>, but having a Result<Option<>>, especially when the error type already has a subtype describing precisely the case of no result found, seems quite redundant to me.

Note that the same change could apply to `bpf_object__find_program_by_title`, I just had a bit of a tunnel vision and didn't notice it until now.

Comment (laarmen) [reply rebpf/src/libbpf.rs:342]: To be more explicit, the `Result<Option<BpfMap>>` wouldn't bother me so much if the `Error::InvalidProgramName` and `Error::InvalidMapName` didn't exist, I just dislike having both ways to report the same condition.

The way I see it, in most uses of this API the map is expected to be there, and its absence marks an actual error somewhere. Using an Option makes this normal use awkward, as you cannot deal with it the same way you'd deal with any unexpected error related to libbpf, you have to make a special case for it.

On the other hand, folding back None into InvalidMapName makes it indeed harder for the caller to pick out this specific case from all the other errors, but it's still possible in a couple of lines.

If you agree with this approach, I'll submit a PR for the find_prog_by_title function.

Comment (uccidibuti) [reply rebpf/src/libbpf.rs:342]: Yes now I unterstand your vision and I agree with your approach, feel you free to do a PR for find_prog_by_title, If you want I can do it too.. 
I'm sorry for my later answer but I was very busy today for works and other reasons, I hope to done a new redirect packet example tonight..

Comment (uccidibuti) [reply rebpf/src/libbpf.rs:342]: Note that I have created a new dev branch, so we can update our commits in dev without update examples evrytime, I will update them when I will merge with branch master..

**Comment by laarmen** [comment]
Hey it's OK not to respond for a while, we all have lives too :-). I'm
lucky to be able to write some of the code for this for work, but I know it
cannot be the case for everyone.

I'll usually use the examples as integration tests of sorts, which means
keeping them in shape. I'm wondering if it would be possible to massage
them into a format that `cargo test` would understand...

**Comment by uccidibuti** [comment]
I think to have a solution to create test for bpf programs, this weekend I will try to implement a cargo test for the bpf examples bpf programs (I am in a strictly quarantine and then I have all weekend available)..

**Comment by uccidibuti** [comment]
I didn't complete bpf programs test, I added PERCPU_ARRAY maps, I was busy last week so what do you think to move rebpf in a organization? Have you some idea from the organization name (maybe rebpf)? So you can add your commit if you want without waiting me (if you agree fell you free to commit when and if you want)..
The only things that I ask is to share my idea to build rebpf api on libbpf safe wrapper and mantain the same name for libbpf C structs/functions but at this point I don't think that this is a problem for you..

About userspace/maps.rs
- Can I move Lookup trait from src/maps.rs to src/userspace/maps.rs?
- What do you think to create Trait for from_obj and extract_info in userspace/maps.rs?

About bpf programs
I had some compile problems writings bpf programs in Rust, but I'm trying to undestand how to write safe bpf programs in Rust..

**Pull request merged**